service/src/config.rs

```diff
@@ -1,61 +1,312 @@
+	}
+
+	/// Returns the trusted worker url that should be addressed by external clients.
+	pub fn trusted_worker_url_external(&self) -> String {
+		match &self.trusted_external_worker_address {
+			Some(external_address) => external_address.to_string(),
+			None => format!("wss://{}:{}", self.worker_ip, self.trusted_worker_port),
+		}
+	}
+
+	pub fn untrusted_worker_url(&self) -> String {
+		format!("{}:{}", self.worker_ip, self.untrusted_worker_port)
+	}
+
+	/// Returns the untrusted worker url that should be addressed by external clients.
+	pub fn untrusted_worker_url_external(&self) -> String {
+		match &self.untrusted_external_worker_address {
+			Some(external_address) => external_address.to_string(),
+			None => format!("ws://{}:{}", self.worker_ip, self.untrusted_worker_port),
+		}
 	}
 
 	pub fn mu_ra_url(&self) -> String {
-		format!("{}:{}", self.worker_ip, self.worker_mu_ra_port)
+		format!("{}:{}", self.worker_ip, self.mu_ra_port)
 	}
 
-	pub fn set_ext_api_url(&mut self, url: String) {
-		self.ext_api_url = Some(url)
+	/// Returns the mutual remote attestion worker url that should be addressed by external workers.
+	pub fn mu_ra_url_external(&self) -> String {
+		match &self.mu_ra_external_address {
+			Some(external_address) => external_address.to_string(),
+			None => format!("{}:{}", self.worker_ip, self.mu_ra_port),
+		}
 	}
 }
 
 impl From<&ArgMatches<'_>> for Config {
 	fn from(m: &ArgMatches<'_>) -> Self {
+		let trusted_port = m.value_of("trusted-worker-port").unwrap_or(DEFAULT_TRUSTED_PORT);
+		let untrusted_port = m.value_of("untrusted-worker-port").unwrap_or(DEFAULT_UNTRUSTED_PORT);
+		let mu_ra_port = m.value_of("mu-ra-port").unwrap_or(DEFAULT_MU_RA_PORT);
+
 		Self::new(
-			m.value_of("node-server").unwrap_or("ws://127.0.0.1").into(),
-			m.value_of("node-port").unwrap_or("9944").into(),
+			m.value_of("node-server").unwrap_or(DEFAULT_NODE_SERVER).into(),
+			m.value_of("node-port").unwrap_or(DEFAULT_NODE_PORT).into(),
 			if m.is_present("ws-external") { "0.0.0.0".into() } else { "127.0.0.1".into() },
-			m.value_of("worker-rpc-port").unwrap_or("2000").into(),
-			m.value_of("mu-ra-port").unwrap_or("3443").into(),
+			m.value_of("trusted-external-address")
+				.map(|url| add_port_if_necessary(url, trusted_port)),
+			trusted_port.to_string(),
+			m.value_of("untrusted-external-address")
+				.map(|url| add_port_if_necessary(url, untrusted_port)),
+			untrusted_port.to_string(),
+			m.value_of("mu-ra-external-address")
+				.map(|url| add_port_if_necessary(url, mu_ra_port)),
+			mu_ra_port.to_string(),
 		)
 	}
 }
+
+fn add_port_if_necessary(url: &str, port: &str) -> String {
+	// [Option("ws(s)"), ip, Option(port)]
+	match url.split(':').count() {
+		3 => url.to_string(),
+		2 => {
+			if url.contains("ws") {
+				// url is of format ws://127.0.0.1, no port added
+				format!("{}:{}", url, port)
+			} else {
+				// url is of format 127.0.0.1:4000, port was added
+				url.to_string()
```

Comment:
This one should be invalid, shouldn't it?

Reply:
mu-ra server actually expects one without. When I tried to add `ws` or `wss` the client failed. But I haven't looked into details there.

Reply:
Oh my gosh, we need to unify the setup here, this is not bearable.